beps/bep_0003.rst

```diff
@@ -87,70 +92,115 @@ announce
 encoded.
 
 info dictionary
-...............
+===============
 
-The ``name`` key maps to a UTF-8 encoded string which is the
-suggested name to save the file (or directory) as. It is purely advisory.
+``name``
+  a UTF-8 encoded string which is the suggested name to save the file (or directory) as.
+  It is purely advisory.
 
-``piece length`` maps to the number of bytes in each piece
-the file is split into. For the purposes of transfer, files are
-split into fixed-size pieces which are all the same length except for
-possibly the last one which may be truncated. ``piece
-length`` is almost always a power of two, most commonly 2 18 =
-256 K (BitTorrent prior to version 3.2 uses 2 20 = 1 M as
-default).
+``piece length``
+  the number of bytes that each logical piece in the peer protocol refers to.
+  I.e. it sets the granularity of ``piece``, ``request``, ``bitfield`` and ``have``
+  messages. It must be a power of two and at least 16KiB.
+  
+  Files are mapped into this piece address space so that each non-empty file starts
+  at a piece boundary and occur in the same order as in the file tree.
+  The last piece of each file may be shorter than the specified piece length.
 
-``pieces`` maps to a string whose length is a multiple of
-20. It is to be subdivided into strings of length 20, each of which is
-the SHA1 hash of the piece at the corresponding index.
+``digest func``
+  the digest used for the calculation of merkle trees and the infohash.
+  Currently valid values are ``sha3-256`` and ``blake2s``.
+  Implementations must reject torrents if they encounter an unknown value.
+  Future revisions may allow additional algorithms if new vulnerabilities are discovered.
 
-There is also a key ``length`` or a key ``files``,
-but not both or neither. If ``length`` is present then the
-download represents a single file, otherwise it represents a set of
-files which go in a directory structure.
+The remaining fields differ depending on whether the torrent represents
+one or more files.
 
-In the single file case, ``length`` maps to the length of
-the file in bytes.
+single-file
+-----------
 
-For the purposes of the other keys, the multi-file case is treated as
-only having a single file by concatenating the files in the order they
-appear in the files list. The files list is the value
-``files`` maps to, and is a list of dictionaries containing
-the following keys:
+``length``
+  Length of the file in bytes.
+  
+``pieces root``
+  The root hash of a merkle tree with a branching factor of 2,
+  constructed from 16KiB blocks of the file.
+  The last block of the file may be smaller than 16KiB.
+  The remaining leaf hashes beyond the end of the file required
+  to construct upper layers of the merkle tree are set to zero.   
+
+multi-file
+----------
+
+``files``
+  is a list of dictionaries which represent files or directories
+  containing additional files or directories.
+  
+  Each dictionary contains
+  
+  ``path``
```

Comment:
> I wonder if there's a more compact (and perhaps simpler) way to represent directory trees.

Yes, using dictionary keys as path elements. It would also have the neat property of enforcing uniqueness.

We can do that, but then we'll have to encode it twice for backwards-compatible torrents. I'm not sure what is better, compromising in the new format or compromising how to achieve backwards compatibility.

> (perhaps I just missed it).

No, it's part of the TODO